actionpack/lib/action_controller/metal/content_security.rb

```diff
@@ -0,0 +1,308 @@
+        define_method(name) do |*sources|
+          if sources.first
+            @directives[directive] = apply_mappings(sources)
+          else
+            @directives.delete(directive)
+          end
+        end
+      end
+
+      def block_all_mixed_content(enabled = true)
+        if enabled
+          @directives["block-all-mixed-content"] = true
+        else
+          @directives.delete("block-all-mixed-content")
+        end
+      end
+
+      def plugin_types(*types)
+        if types.first
+          @directives["plugin-types"] = types
+        else
+          @directives.delete("plugin-types")
+        end
+      end
+
+      def report_only!(uri:)
+        @report_only = true
+        @directives["report-uri"] = [uri]
+      end
+
+      def report_uri(uri)
+        @directives["report-uri"] = [uri]
+      end
+
+      def require_sri_for(*types)
+        if types.first
+          @directives["require-sri-for"] = types
+        else
+          @directives.delete("require-sri-for")
+        end
+      end
+
+      def sandbox(*values)
+        if values.empty?
+          @directives["sandbox"] = true
+        elsif values.first
+          @directives["sandbox"] = values
+        else
+          @directives.delete("sandbox")
+        end
+      end
+
+      def upgrade_insecure_requests(enabled = true)
+        if enabled
+          @directives["upgrade-insecure-requests"] = true
+        else
+          @directives.delete("upgrade-insecure-requests")
+        end
+      end
+
+      def header_name
+        if @report_only
+          "Content-Security-Policy-Report-Only"
+        else
+          "Content-Security-Policy"
+        end
+      end
+
+      def header_value(context = nil)
+        build_directives(context).compact.join("; ") + ";"
+      end
+
+      private
+        def copy_directives(directives)
+          {}.tap do |copy|
+            directives.each do |directive, sources|
+              copy[directive] = sources.map(&:dup)
+            end
+          end
+        end
+
+        def apply_mappings(sources)
+          sources.map do |source|
+            case source
+            when Symbol
+              apply_mapping(source)
+            when String, Proc
+              source
+            else
+              raise ArgumentError, "Invalid content security policy source: #{source.inspect}"
+            end
+          end
+        end
+
+        def apply_mapping(source)
+          if MAPPINGS.key?(source)
+            MAPPINGS.fetch(source)
+          else
+            raise ArgumentError, "Unknown content security policy source mapping: #{source.inspect}"
+          end
+        end
+
+        def build_directives(context)
+          @directives.map do |directive, sources|
+            if sources.is_a?(Array)
+              "#{directive} #{build_directive(sources, context).join(' ')}"
+            elsif sources
+              directive
+            else
+              nil
```

Comment:
Appears that a directive sources will never be `nil`, should just be absent. Raise here?

Reply:
This is when a directive is set to `false` - there's a `compact` operation that removes these nil values in `header_value`:

``` ruby
def header_value(context = nil)
  build_directives(context).compact.join("; ") + ";"
end
```